presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java

```diff
@@ -771,78 +779,192 @@ public void end()
+
+            boolean outOfOrder = false;
+            ImmutableList.Builder<BlockFieldConverter> builder = ImmutableList.builder();
+            for (int i = 0, prevFieldIndex = -1; i < fieldCount; i++) {
                 parquet.schema.Type fieldType = fieldTypes.get(i);
-                converters.add(createConverter(prestoTypeParameters.get(i), columnName + "." + fieldType.getName(), fieldType, i));
+                Pair pair = fieldTypeMap.get(fieldType.getName().toLowerCase());
+                if (pair == null) {
+                    log.info("Schema mismatch, metastore schema for row column %s doesn't have field %s but parquet schema has it",
+                            columnName,
+                            fieldType.getName());
+                    builder.add(fieldType.isPrimitive() ? ParquetDummyPrimitiveConverter.converter : new ParquetDummyGroupConverter(fieldType));
+                }
+                else {
+                    if (prevFieldIndex > pair.index) {
+                        outOfOrder = true;
+                    }
+                    builder.add(createConverter(pair.type, columnName + "." + fieldType.getName(), fieldType, pair.index, true));
+                    prevFieldIndex = pair.index;
+                }
             }
-            this.converters = converters.build();
+            // When reading parquet files, we need to pass converters and parquet types in parquet schema fields order.
+            // After getting the parquet values, we need to build presto blocks in metastore schema fields order.
+            List<BlockFieldConverter> converters = builder.build();
+            if (outOfOrder) {
+                log.info("Schema mismatch, struct field order for metastore schema for row column %s is different from that of parquet schema",
+                        columnName);
+                ImmutableList.Builder<Type> types = ImmutableList.builder();
+                converterIndexes = new int[fieldCount];
+                for (int i = 0, tmpIndex = 0; i < fieldCount; i++) {
+                    BlockFieldConverter converter = converters.get(i);
+                    converterIndexes[i] = converter.getFieldIndex();
+                    if (converterIndexes[i] < 0) {
+                        continue;
+                    }
+                    converter.setFieldIndex(tmpIndex++);
+                    types.add(converter.getType());
+                }
+                outOfOrderFieldTypes = types.build();
+            }
+            return converters;
+        }
+
+        public ParquetStructConverter(Type prestoType, String columnName, GroupType entryType, int fieldIndex, boolean useNames)
+        {
+            checkArgument(ROW.equals(prestoType.getTypeSignature().getBase()));
+            List<parquet.schema.Type> fieldTypes = entryType.getFields();
+
+            if (useNames) {
+                this.converters = createConvertersByName(columnName, ((RowType) prestoType).getFields(), fieldTypes);
```

Comment:
Ya, you should be able to just use the `TypeSignature` all the names will be in there